libcaf_net/test/net/web_socket/framing.cpp

```diff
@@ -0,0 +1,184 @@
+  }
+};
+
+} // namespace
+
+BEGIN_FIXTURE_SCOPE(fixture)
+
+SCENARIO("the client sends a ping and receives a pong response") {
+  GIVEN("a valid WebSocket connection") {
+    std::vector<std::byte> ping_frame;
+    std::vector<std::byte> pong_frame;
+    CHECK_EQ(transport->start(nullptr), error{});
+    transport->configure_read(caf::net::receive_policy::up_to(2048u));
+    WHEN("the client sends a ping with no data") {
+      auto data = make_test_data(0);
+      detail::rfc6455::assemble_frame(detail::rfc6455::ping, 0x0, data,
+                                      ping_frame);
+      transport->push(ping_frame);
+      CHECK_EQ(transport->handle_input(),
+               static_cast<ptrdiff_t>(ping_frame.size()));
+      THEN("the client receives an empty pong") {
+        detail::rfc6455::assemble_frame(detail::rfc6455::pong, 0x0, data,
+                                        pong_frame);
+        CHECK_EQ(transport->output_buffer(), pong_frame);
+      }
+    }
+    transport->output_buffer().clear();
+    WHEN("the client sends a ping with some data") {
+      auto data = make_test_data(40);
+      detail::rfc6455::assemble_frame(detail::rfc6455::ping, 0x0, data,
+                                      ping_frame);
+      transport->push(ping_frame);
+      CHECK_EQ(transport->handle_input(),
+               static_cast<ptrdiff_t>(ping_frame.size()));
+      THEN("the client receives a pong containing the data echoed back") {
+        detail::rfc6455::assemble_frame(detail::rfc6455::pong, 0x0, data,
+                                        pong_frame);
+        CHECK_EQ(transport->output_buffer(), pong_frame);
+      }
+    }
+    transport->output_buffer().clear();
+    WHEN("the client sends a ping with the maximum amount of data") {
+      auto data = make_test_data(125);
+      detail::rfc6455::assemble_frame(detail::rfc6455::ping, 0x0, data,
+                                      ping_frame);
+      transport->push(ping_frame);
+      CHECK_EQ(transport->handle_input(),
+               static_cast<ptrdiff_t>(ping_frame.size()));
+      THEN("the client receives a pong containing the data echoed back") {
+        detail::rfc6455::assemble_frame(detail::rfc6455::pong, 0x0, data,
+                                        pong_frame);
+        CHECK_EQ(transport->output_buffer(), pong_frame);
+      }
+    }
+  }
+}
+
+TEST_CASE("calling shutdown with protocol_error sets status in close header") {
+  CHECK_EQ(transport->start(nullptr), error{});
+  transport->configure_read(caf::net::receive_policy::up_to(2048u));
+  uut->shutdown(make_error(sec::protocol_error));
+  detail::rfc6455::header hdr;
+  detail::rfc6455::decode_header(transport->output_buffer(), hdr);
+  CHECK_EQ(hdr.opcode, detail::rfc6455::connection_close);
+  CHECK(hdr.payload_len >= 2);
+  auto error_code = (static_cast<uint16_t>(transport->output_buffer()[2]) << 8)
+                    + static_cast<uint16_t>(transport->output_buffer()[3]);
```

Comment:
1) Since we only use this to compare values, I wouldn't bother with the uint16 type.
2) I think using `vector::at` here makes the code a bit easier to read. The bound checks are also nice to have in tests to get exceptions rather than ASAN traces.

```suggestion
  auto error_code = (transport->output_buffer().at(2) << 8)
                    + transport->output_buffer().at(3);
```